### DISCUSSION TITLE
How to connect locally?

### DISCUSSION THREAD

**Discussion by ghost** [mbrubeck/agate#106] [Q&A] [answered]
I've started a server with a test page on a raspberry pi. I have a client (Elaho) on my phone. I've tried `gemini://<ip address>:1965` but get the error:

The operation couldn’t be completed. (Network.NWError error 0.)

I followed these instructions: http://techrights.org/2021/02/23/gemini-server-with-agate/

**Comment by Johann150** [comment]
`gemini://:1965` is not a valid URL so I am surprised it did something at all. You will have to use a domain name or an IP adress in the URL to tell your browser which server to contact.

As I understand the server and the browser are on different devices. You will have to find out the IP address of the Raspberry Pi. You should be able to do that by typing `ip address` into a terminal emulator on the Raspberry Pi. The output should look something like this:

```sh
$ ip address
1: lo: <LOOPBACK, UP, LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: enp0s3: <BROADCAST, MULTICAST, UP, LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 01:23:45:67:89:ab brd ff:ff:ff:ff:ff:ff
    inet 192.168.0.59/24 scope 192.168.0.255 scope link enp0s3:avahi
       valid_lft forever preferred_lft forever
    inet6 fe80::a00:27ff:fe36:19bf/64 scope link
       valid_lft forever preferred_lft forever
```
There are two network adapters, one is the loopback (named `lo` or similar and it says `LOOPBACK` next to the name). The other one is the interesting one. The line that starts with `inet` contains the IPv4 address, in this case `192.168.0.59`; disregard the `/24` part. If you don't have that line you should have a line that starts with `inet6` and having the IPv6 address in it.

With the IP address of your Raspberry Pi you can then go to `gemini://192.168.0.59/` in your Elaho browser. Of course swap my example IP address for your real one. Note that if you run Agate on the default port (1965), you do not have to specify the port in the URL.

**Reply by ghost** [reply]
Thanks Johann. I didn't realise GitHub stripped away my angle brackets and text. I'm actually using:

`gemini://<my ip>:1965`. In my case 192.168.0.101

The terminal where I started agate outputs the following each time Elaho tries to connect:

[2021-11-15T16:42:43Z WARN  agate] 192.168.0.101:1965 - "" 00 "TLS error" error:unexpected error: no server certificate chain resolved

I did try to generate one in .certificates using the following:

`openssl req -new -newkey rsa:4096 -x509 -sha512 -days 3650 -nodes -out Certificate.crt -keyout PrivateKey.key`

Is that incorrect? I'm just looking for a self-signed setup so I can get back to my aim of generating gemini pages dynamically from a separate process.

**Comment by Johann150** [comment, marked as answer]
If you use an IP address, Elaho can not send TLS Server Name Identification (SNI) which is required by the Gemini protocol. That actually means I was wrong in my other answer (which is why I have hidden it). With Gemini it is impossible to use an IP address for the Gemini protocol because SNI is *required*.

Instead you could try to set up your own local DNS server, or simpler manipulate your `/etc/hosts` file, or simpler yet install a Gemini client on the same device as you are running Agate on to make use of the `localhost` hostname that exists by default. This name can also be used for certificates. 

Since the article you linked was written for Agate major version 2 while the current major version is 3, you will probably find the added `--hostname` flag interesting: If you use `--hostname localhost`, Agate will create the necessary certificate to serve as that hostname by itself and you won't need to use `openssl`.

**Comment by ghost** [comment]
Thank you again! That worked. I started the server with

`$ bin/agate --content gemini/ --hostname localhost --lang enUS`

and ran this client for raspberry pi using gemini://localhost as the URL. ( [https://github.com/a-h/min](https://github.com/a-h/min) )

If I do want to generate the certificates for the wider network in the future, is it a subject independent of this project? Because if it's an independent procedure I should search online instead of asking here.

**Reply by Johann150** [reply]
You can use the `--hostname` flag with other hostnames too.
